### PR TITLE
Truffle Decoupling

### PR DESCRIPTION
**Pull request by nlordell**
This PR decouples the deployment futures from truffle. This accomplishes two things:
- Sets us up for generating code from non-truffle sources (for example #112)
- Gives us more control over how the generated code (for example #100) 

### Test Plan

CI - This is mostly a refactoring and adding new abstractions

### CODE REVIEW

File: src/contract.rs [hunk 1]
```diff
@@ -138,6 +140,33 @@ impl<T: Transport> Instance<T> {
     }
 }
 
+/// `Deployments` implementation for an `Instance`. This type is not intended to
+/// be used directly but rather through the `Instance::deployed` API.
+#[derive(Debug, Clone)]
+pub struct Networks {
```
Comment (fleupold) [inline src/contract.rs]: I find this name a bit confusing. Why would a network have a field called ABI? Maybe Deployments is a better name?

File: src/contract.rs [hunk 2]
```diff
@@ -6,14 +6,16 @@ mod deploy;
 mod method;
 
 use crate::errors::{DeployError, LinkError};
-use crate::truffle::abi::Result as AbiResult;
-use crate::truffle::{Abi, Artifact, Bytecode};
+use ethcontract_common::abi::Result as AbiResult;
+use ethcontract_common::truffle::Network;
```
Comment (fleupold) [inline src/contract.rs:10]: Is this still truffle specific or should this also be lifted into common?

**Review by fleupold — APPROVED** [review]
Looks good, although I don't 100% understand how you broke the truffle dependency. You basically lifted the ABI module into common namespace, right?

File: src/contract.rs (hunk 2 above)
Comment (nlordell) [reply src/contract.rs:10]: It is truffle specific.

**Comment by nlordell** [comment]
@fleupold So, we are still very much dependent on Truffle artifacts for generating code. The important decoupling is that deployment (so `DeployedFuture` and `Deploy{Builder,Future}`) no longer depend on a truffle `Artifact`. This allows the generated code to slowly add support for things outside of truffle for example:
- Custom networks definitions (for the dev networks in dex-services for example so we can use pre-built contracts from the npm package).
- Support for other formats like ethPM and maybe getting verified contracts directly from etherscan.io

While the `Instance` type still depends on Truffle artifacts, the building blocks for it no longer do so we can start generating code that doesn't as well.

File: src/contract.rs (hunk 1 above)
Comment (nlordell) [reply src/contract.rs]: Definitely a better name!

**Pull request merged**